pkg/virt-controller/watch/ovm.go

```diff
@@ -0,0 +1,695 @@
+	"kubevirt.io/kubevirt/pkg/controller"
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+func NewOVMController(vmInformer cache.SharedIndexInformer, vmOVMInformer cache.SharedIndexInformer, recorder record.EventRecorder, clientset kubecli.KubevirtClient) *OVMController {
+
+	c := &OVMController{
+		Queue:         workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		vmInformer:    vmInformer,
+		vmOVMInformer: vmOVMInformer,
+		recorder:      recorder,
+		clientset:     clientset,
+		expectations:  controller.NewUIDTrackingControllerExpectations(controller.NewControllerExpectations()),
+	}
+
+	c.vmOVMInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addOvm,
+		DeleteFunc: c.deleteOvm,
+		UpdateFunc: c.updateOvm,
+	})
+
+	c.vmInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addVirtualMachine,
+		DeleteFunc: c.deleteVirtualMachine,
+		UpdateFunc: c.updateVirtualMachine,
+	})
+
+	return c
+}
+
+type OVMController struct {
+	clientset     kubecli.KubevirtClient
+	Queue         workqueue.RateLimitingInterface
+	vmInformer    cache.SharedIndexInformer
+	vmOVMInformer cache.SharedIndexInformer
+	recorder      record.EventRecorder
+	expectations  *controller.UIDTrackingControllerExpectations
+}
+
+func (c *OVMController) Run(threadiness int, stopCh chan struct{}) {
+	defer controller.HandlePanic()
+	defer c.Queue.ShutDown()
+	log.Log.Info("Starting OfflineVirtualMachine controller.")
+
+	// Wait for cache sync before we start the controller
+	cache.WaitForCacheSync(stopCh, c.vmInformer.HasSynced, c.vmOVMInformer.HasSynced)
+
+	// Start the actual work
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(c.runWorker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	log.Log.Info("Stopping OfflineVirtualMachine controller.")
+}
+
+func (c *OVMController) runWorker() {
+	for c.Execute() {
+	}
+}
+
+func (c *OVMController) Execute() bool {
+	key, quit := c.Queue.Get()
+	if quit {
+		return false
+	}
+	defer c.Queue.Done(key)
+	if err := c.execute(key.(string)); err != nil {
+		log.Log.Reason(err).Infof("re-enqueuing OfflineVirtualMachine %v", key)
+		c.Queue.AddRateLimited(key)
+	} else {
+		log.Log.V(4).Infof("processed OfflineVirtualMachine %v", key)
+		c.Queue.Forget(key)
+	}
+	return true
+}
+
+func (c *OVMController) execute(key string) error {
+
+	obj, exists, err := c.vmOVMInformer.GetStore().GetByKey(key)
+	if err != nil {
+		return nil
+	}
+	if !exists {
+		// nothing we need to do. It should always be possible to re-create this type of controller
+		c.expectations.DeleteExpectations(key)
+		return nil
+	}
+	OVM := obj.(*virtv1.OfflineVirtualMachine)
+
+	logger := log.Log.Object(OVM)
+
+	logger.Infof("Working on OVM: %s", OVM.Name)
+
+	//TODO default rs if necessary, the aggregated apiserver will do that in the future
+	if OVM.Spec.Template == nil || OVM.Spec.Selector == nil || len(OVM.Spec.Template.ObjectMeta.Labels) == 0 {
```

Comment:
We can remove the label matching from the controller and use the vm name instead.